src/groups.py

```diff
@@ -0,0 +1,380 @@
+
+        Args:
+            charm: the charm for which this relation is provided
+        """
+        self.charm = charm
+
+    @property
+    def _auth(self):
+        """Return the authentication credentials."""
+        admin_password = self.charm.config["ranger-admin-password"]
+        auth = (ADMIN_USER, admin_password)
+        return auth
+
+    @log_event_handler(logger)
+    def _handle_synchronize_file(self, event):
+        """Synchronize the data from the config file with the Ranger API.
+
+        Args:
+            event: The config changed event that triggered the synchronization.
+        """
+        if not self.charm.unit.status == ActiveStatus("Status check: UP"):
+            logger.debug("Status not active, event deferred.")
+            event.defer()
+            return
+
+        if not self.charm.unit.is_leader():
+            return
+
+        config_data = self._read_file()
+        for key in config_data:
+            data = next(iter(config_data.values()))
+            relation_data = copy.deepcopy(data)
+            self._synchronize(key, data, event)
+            self._add_to_relations(key, relation_data)
+
+    def _read_file(self):
+        """Read the user-group-configuration from the charm config.
+
+        Returns:
+            Data read from the configuration file.
+        """
+        config = self.charm.config["user-group-configuration"]
+        data = yaml.safe_load(config)
+        return data
+
+    def _synchronize(self, key, data, event):
+        """Synchronize data with the Ranger API.
+
+        Args:
+            key: relation_id from configuration file.
+            data: Data to synchronize with the Ranger API.
+            event: configuration changed event.
+        """
+        try:
+            self._sync(data["groups"], "group")
+            self._sync(data["users"], "user")
+            self._sync(data["memberships"], "membership")
+            logger.info(f"Synchronized users and groups for {key}")
+        except RangerServiceException:
+            logger.exception(
+                f"A Ranger Service Exception has occurred while attempting to sync {key}."
+            )
+            event.defer()
+
+    @handle_service_error
+    def _sync(self, config, member_type):
+        """Synchronize apply values with the Ranger API.
+
+        Args:
+            config: Values to synchronize.
+            member_type: The type of Ranger member (group, user or membership).
+        """
+        # Get existing values
+        existing = self._get_existing_values(member_type)
+
+        # Get values to apply
+        apply = self._transform_apply_values(config, member_type)
+
+        # Create members
+        to_create = [item for item in apply if item not in existing]
+        for value in to_create:
+            if member_type == "membership":
+                fields = value
+            if member_type in ["user", "group"]:
+                fields = [
```

Comment:
[Style] It might make sense to use `next` with a generator here. Though I am reading that it is not always recommended.